onnxscript/ir/_core.py

```diff
@@ -436,6 +437,51 @@ def __ge__(self, other: object) -> bool:
             raise TypeError(f"Expected other to be Dimension or int, got {type(other)}")
         return int(self) >= int(other)
 
+    def __neg__(self) -> int:
+        return -int(self)
+
+    def __add__(self, other):
```

Comment:
My intuition is that this is not worth adding. It is likely to be misused. Logically, what is useful (eg., in shape-inference context) is something that handles the symbolic case also ... even if it is to just produce an "unknown" Dimension value (represented as None). Eg.: something like [this](https://github.com/onnx/onnx/blob/c459890aa266d74ca612f334d0f3e869dcdbb597/onnx/defs/shape_inference.h#L175) for multiplication.

Reply:
In building the graph for torchlib, I have seen traced only functions doing `shape[0] // 2`, which prompted this update

Reply:
We can also return dim.value when a user is indexing on shape; and provide another api for symbolic dimensions

Reply:
But if we use `.dims` the getter `shape.dims[...]` will not be symmetrical with the setter `shape[...]`

Reply:
(i) We can't assume self is an int here. If it is not, there is a way to return a valid Dimension.

(ii) Wrt to particular use-case you mention: if a traceonly function is doing `shape[0] // 2`, then what do we want to do? The general solution is to emit ONNX nodes/ops that represent this computation (with a statically known value being a special case). 

Basically, when `shape` itself is computed, it should NOT be an instance of `ir.Shape`. It should be an instance of `ir.Value` ... and `shape[0]` should also be an instance of `ir.Value` (not an `ir.Dimension`). Or, am I mistaken?

Reply:
We can flatten the Shape class a little. Use int for known dimensions, and use Dim(str|None) for symbolic dimensions (this way two strings are not accidentally compared).

We can then store the denotations in the Shape class.

Reply:
So the use case above is for static evaluation; we didn't want to emit dynamic code in that particular instance of the trace function. The result of `shape[0] // 2` was used as a constant

Reply:
When a user wishes to have a dynamic shape, I would expect they do `op.Shape(<ir.Value>)`; whereas `ir.Value.shape` returns static shape information that is part of its type.

Reply:
I have reverted this part of the change and will work on it in a separate PR.

Reply:
https://github.com/microsoft/onnxscript/pull/1399